File: frontend/src/lib/demo/analytics.ts
/**
 * The demo's port of `app/services/analytics.py`.
 *
 * Same rule as the server's: everything here is a *view*. It ranks, filters and
 * compares answers the ported decision, market and submission engines already
 * gave, and never re-derives one of them. `make parity` runs both this file and
 * the Python it mirrors over identical cases and compares every field, so a
 * divergence fails the build rather than quietly shipping to Pages.
 */

import type {
  BlockStatus,
  CollectionDecisions,
  CollectionFilter,
  Decision,
  FilterResult,
  GradedCardResult,
  Opportunity,
  RankedOpportunities,
  SellingCandidate,
  SellingQueue,
  Submission,
  SubmissionReturn,
  SubmissionReturns,
} from '@/lib/types'

/** Verdicts that mean money would be spent on grading. */
const ACTIONABLE = new Set<Decision>(['grade', 'grade_if_batch_filled'])

/** Verdicts that mean the card is not going to a grader. */
const SELLABLE = new Set<Decision>(['sell_raw', 'keep_raw', 'do_not_grade'])

/**
 * Directions that mean the price is going the wrong way. `insufficient_data` is
 * deliberately absent: not knowing is not the same as falling.
 */
const FALLING = new Set(['down', 'strong_down'])

export function rankOpportunities(decisions: CollectionDecisions): RankedOpportunities {
  const actionable = decisions.opportunities.filter((item) => ACTIONABLE.has(item.decision))
  return {
    currency: decisions.currency,
    analysed: decisions.analysed,
    total_cards: decisions.total_cards,
    actionable: actionable.length,
    expected_profit: decisions.expected_profit,
    total_grading_cost: decisions.total_grading_cost,
    items: actionable,
    status: decisions.status,
    reason: actionable.length
      ? `${actionable.length} of ${decisions.analysed} analysed card(s) are worth grading. ${
          decisions.reason ?? ''
        }`.trim()
      : decisions.reason,
  }
}

/**
 * What to *ask*, which is not what it will *fetch*.
 *
 * A listing price is a negotiating position: you list above the realistic sale
 * and expect to come down. How far above is a liquidity question — a card that
 * trades weekly can be listed near its median and still move, while one that
 * trades twice a year needs room to be haggled down and time to find its buyer.
 *
 * The upper quartile of actual sales is the ceiling, because asking more than
 * anyone has recently paid is how a listing sits unsold for a year.
 */
export function suggestedListingMinor(
  realisticMinor: number | null,
  highQuartileMinor: number | null,
  liquidityScore: number | null,
): [number | null, string | null] {
  if (realisticMinor === null) return [null, null]

  let markup: number
  let basis: string
  if (liquidityScore === null) {
    markup = 0.1
    basis = '10% above the realistic sale price — no liquidity reading to judge by.'
  } else if (liquidityScore >= 7) {
    markup = 0.05
    basis = '5% above the realistic sale price. It trades often, so it does not need room.'
  } else if (liquidityScore >= 4) {
    markup = 0.1
    basis = '10% above the realistic sale price, leaving a little room to negotiate.'
  } else {
    markup = 0.18
    basis =
      '18% above the realistic sale price. It trades rarely, so it needs room to be haggled ' +
      'down and time to find a buyer.'
  }

  let asking = Math.round(realisticMinor * (1 + markup))

  // Never below the realistic sale price: the cap exists to stop you asking
  // more than the market pays, not to talk you into asking less than it does.
  if (highQuartileMinor && realisticMinor < highQuartileMinor && highQuartileMinor < asking) {
    asking = highQuartileMinor
    basis =
      'Capped at the upper quartile of recent sales — asking more than anyone has recently ' +
      'paid is how a listing sits unsold.'
  }
  return [asking, basis]
}

export interface QueueInput {
  /** The card's raw price row, if it has one. */
  realisticMinor: number | null
  highQuartileMinor: number | null
  confidence: string
  liquidityScore: number | null
  liquidityBand: string | null
  daysSinceLastSale: number | null
  trendDirection: string | null
  purchasePrice: number | null
}

export function buildSellingQueue(
  decisions: CollectionDecisions,
  lookup: (cardId: string) => QueueInput,
): SellingQueue {
  const result: SellingQueue = {
    currency: decisions.currency,
    analysed: decisions.analysed,
    total_cards: decisions.total_cards,
    total_net_proceeds: null,
    items: [],
    status: 'ok',
    reason: null,
    notes: [],
  }

  const sellable = decisions.opportunities.filter((item) => SELLABLE.has(item.decision))
  if (!sellable.length) {
    result.status = 'insufficient_data'
    result.reason = decisions.analysed
      ? 'Nothing in the analysed cards is better off sold raw right now.'
      : decisions.reason
    return result
  }

  let netTotal = 0
  let counted = 0
  for (const item of sellable) {
    const market = lookup(item.card_id)
    const candidate: SellingCandidate = {
      card_id: item.card_id,
      name: item.name,
      set_label: item.set_label,
      decision: item.decision,
      realistic_sale: null,
      net_proceeds: item.net_raw_alternative,
      suggested_listing: null,
      listing_basis: null,
      liquidity_score: market.liquidityScore,
      liquidity_band: market.liquidityBand,
      days_since_last_sale: market.daysSinceLastSale,
      trend_direction: market.trendDirection,
      confidence: market.confidence as SellingCandidate['confidence'],
      purchase_price: market.purchasePrice,
      gain_vs_purchase: null,
      blockers: [],
    }

    if (market.realisticMinor !== null) {
      candidate.realistic_sale = market.realisticMinor / 100
      const [asking, basis] = suggestedListingMinor(
        market.realisticMinor,
        market.highQuartileMinor,
        market.liquidityScore,
      )
      candidate.suggested_listing = asking === null ? null : asking / 100
      candidate.listing_basis = basis
    } else {
      candidate.blockers.push(
        'No raw sales stored, so there is nothing to price a listing against.',
      )
    }

    if (candidate.net_proceeds !== null && market.purchasePrice !== null) {
      candidate.gain_vs_purchase = round2(candidate.net_proceeds - market.purchasePrice)
    }
    if (candidate.net_proceeds !== null) {
      netTotal += Math.round(candidate.net_proceeds * 100)
      counted += 1
    }
    result.items.push(candidate)
  }

  // Most valuable first: this is a to-do list, and the biggest cheque is the
  // one worth writing the listing for tonight.
  result.items.sort((a, b) => (b.net_proceeds ?? 0) - (a.net_proceeds ?? 0))
  if (counted) result.total_net_proceeds = round2(netTotal / 100)

  const unpriced = result.items.filter((item) => item.suggested_listing === null)
  if (unpriced.length) {
    result.status = 'partial'
    result.notes.push(
      `${unpriced.length} card(s) have no raw sales to price against, so no listing price is ` +
        'suggested for them.',
    )
    result.reason = result.notes[0]
  }
  return result
}

/**
 * The cuts offered in the UI. Each is a predicate over figures the decision and
 * market engines already produced — never a fresh definition of the same idea.
 */
export const FILTERS: CollectionFilter[] = [
  { key: 'grade_now', label: 'Grade now', description: 'Clears your bar on its own today.' },
  {
    key: 'grade_if_batch_filled',
    label: 'Grade in a batch',
    description: 'Worth grading, but only once a submission is full.',
  },
  { key: 'sell_raw', label: 'Sell raw', description: 'Better off sold as it is.' },
  {
    key: 'hold',
    label: 'Hold',
    description: 'Grading does not pay yet, but the market is rising.',
  },
  {
    key: 'high_upside',
    label: 'High upside',
    description: 'The good outcomes are worth a lot more than the bad ones.',
  },
  {
    key: 'high_risk',
    label: 'High risk',
    description: 'A real chance of losing money against selling it raw.',
  },
  {
    key: 'low_liquidity',
    label: 'Hard to sell',
    description: 'Trades below the minimum liquidity you set, so any plan for it will be slow.',
  },
  {
    key: 'declining',
    label: 'Declining',
    description: 'Prices are falling — waiting is costing you.',
  },
  {
    key: 'needs_data',
    label: 'Needs data',
    description: 'Cannot be decided until it has more behind it.',
  },
]

export function applyFilter(
  key: string,
  decisions: CollectionDecisions,
  minimumLiquidity: number,
): FilterResult {
  const definition = FILTERS.find((item) => item.key === key)
  if (!definition) throw new Error(`'${key}' is not a collection filter.`)

  const matched = decisions.opportunities.filter((item) =>
    matches(key, item, minimumLiquidity),
  )
  const result: FilterResult = {
    key,
    label: definition.label,
    description: definition.description,
    currency: decisions.currency,
    matched: matched.length,
    analysed: decisions.analysed,
    total_cards: decisions.total_cards,
    // Cards the sweep never reached plus those it could not decide: both are
    // unanswered, and lumping them into "does not match" would be a lie.
    unclassified:
      key === 'needs_data'
        ? 0
        : decisions.skipped_not_ready + (decisions.counts.insufficient_data ?? 0),
    card_ids: matched.map((item) => item.card_id),
    items: matched,
    status: 'ok' as BlockStatus,
    reason: null,
  }

  if (!decisions.analysed) {
    result.status = 'insufficient_data'
    result.reason = decisions.reason
  } else if (result.unclassified) {
    result.status = 'partial'
    result.reason =
      `${result.unclassified} card(s) could not be decided, so they were not tested against ` +
      'this filter.'
  }
  return result
}

/**
 * Whether one card falls into a named cut.
 *
 * Every branch reads a figure the decision engine produced. Where a figure is
 * unknown the card does **not** match: an unknown risk is not a low risk, and a
 * card with no trend behind it is not a falling one.
 */
function matches(key: string, item: Opportunity, minimumLiquidity: number): boolean {
  if (key === 'grade_now') return item.decision === 'grade'
  if (key === 'grade_if_batch_filled') return item.decision === 'grade_if_batch_filled'
  if (key === 'sell_raw') return item.decision === 'sell_raw'
  if (key === 'hold') return item.decision === 'hold'
  if (key === 'needs_data') return item.decision === 'insufficient_data'

  if (key === 'high_upside') {
    // Judged against what selling it raw would net, so "high" means high
    // relative to the alternative rather than high in absolute pounds.
    if (item.expected_profit === null || !item.net_raw_alternative) return false
    return item.expected_profit >= item.net_raw_alternative * 0.5
  }

  if (key === 'high_risk') {
    if (item.probability_of_profit === null) return false
    return item.probability_of_profit < 0.5 && ACTIONABLE.has(item.decision)
  }

  if (key === 'low_liquidity') {
    if (item.liquidity_score === null) return false
    return item.liquidity_score < minimumLiquidity
  }

  if (key === 'declining') {
    return item.trend_direction !== null && FALLING.has(item.trend_direction)
  }

  return false
}

/* --- Submission returns ---------------------------------------------------- */

export interface ReturnInput {
  submission: Submission
  /** What one slab is worth at its actual grade, or null when nobody has sold one. */
  gradedValueMinor: (cardId: string, grade: number) => number | null
  /** What that gross would net after fees and postage. */
  netOf: (grossMinor: number) => number | null
}

/**
 * What the parcels you have sent actually returned.
 *
 * Only submissions that have come back can be scored. The rest are counted and
 * reported rather than averaged in at zero, which would make every open
 * submission look like a loss.
 */
export function buildSubmissionReturns(
  currency: string,
  inputs: ReturnInput[],
): SubmissionReturns {
  const result: SubmissionReturns = {
    currency,
    submissions: [],
    scored: 0,
    awaiting: 0,
    total_cost: null,
    total_profit: null,
    roi_pct: null,
    status: 'ok',
    reason: null,
  }

  let costMinor = 0
  let profitMinor = 0

  for (const { submission, gradedValueMinor, netOf } of inputs) {
    const entry: SubmissionReturn = {
      submission_id: submission.id,
      reference: submission.reference,
      company_code: submission.company_code,
      status: submission.status,
      returned_at: submission.returned_at,
      card_count: submission.card_count,
      graded_count: 0,
      total_cost: submission.total_cost,
      total_value: null,
      total_profit: null,
      roi_pct: null,
      mean_surprise: null,
      cards: [],
      status_note: null,
    }

    const graded = submission.cards.filter((row) => row.actual_grade !== null)
    entry.graded_count = graded.length

    if (!graded.length) {
      result.awaiting += 1
      entry.status_note =
        'No grades recorded yet, so there is nothing to score. Record the grades when the ' +
        'parcel comes back.'
      result.submissions.push(entry)
      continue
    }

    let valueMinor = 0
    let entryCostMinor = 0
    const surprises: number[] = []

    for (const row of graded) {
      const lineCostMinor = Math.round((row.total_cost ?? 0) * 100)
      const card: GradedCardResult = {
        card_id: row.card_id,
        name: row.name,
        predicted_grade: row.predicted_grade,
        actual_grade: row.actual_grade,
        surprise: null,
        cost: row.total_cost,
        graded_value: null,
        net_if_sold: null,
        profit: null,
        blockers: [],
      }
      if (row.predicted_grade !== null && row.actual_grade !== null) {
        card.surprise = round2(row.actual_grade - row.predicted_grade)
        surprises.push(card.surprise)
      }

      const worth = gradedValueMinor(row.card_id, row.actual_grade!)
      if (worth === null) {
        card.blockers.push(
          `No ${submission.company_code} sales stored at grade ${formatGrade(row.actual_grade!)}, ` +
            'so this slab cannot be valued.',
        )
      } else {
        card.graded_value = worth / 100
        const net = netOf(worth)
        if (net !== null) {
          card.net_if_sold = net / 100
          valueMinor += net
          card.profit = round2((net - lineCostMinor) / 100)
        }
      }
      entryCostMinor += lineCostMinor
      entry.cards.push(card)
    }

    entry.total_value = valueMinor ? round2(valueMinor / 100) : null
    if (entryCostMinor && valueMinor) {
      entry.total_profit = round2((valueMinor - entryCostMinor) / 100)
      entry.roi_pct = round1(((valueMinor - entryCostMinor) / entryCostMinor) * 100)
      costMinor += entryCostMinor
      profitMinor += valueMinor - entryCostMinor
    }
    if (surprises.length) {
      entry.mean_surprise = round2(surprises.reduce((a, b) => a + b, 0) / surprises.length)
    }

    // A slab nobody has sold still cost money to grade. Its fee is in the total
    // and its value is not, so the return is a floor rather than an estimate.
    const unvalued = entry.cards.filter((row) => row.graded_value === null).length
    if (unvalued && entry.roi_pct !== null) {
      entry.status_note =
        `${unvalued} of ${entry.graded_count} graded card(s) have no sales at that grade, so ` +
        'they cost money here but add no value. The return is a floor, not an estimate.'
    }

    result.scored += 1
    result.submissions.push(entry)
  }

  if (costMinor) {
    result.total_cost = round2(costMinor / 100)
    result.total_profit = round2(profitMinor / 100)
    result.roi_pct = round1((profitMinor / costMinor) * 100)
  }

  if (!inputs.length) {
    result.status = 'insufficient_data'
    result.reason = 'No submissions yet, so there is nothing to score.'
  } else if (!result.scored) {
    result.status = 'insufficient_data'
    result.reason =
      `${result.awaiting} submission(s) are still out. Record the grades when they come back ` +
      'and this becomes a real return.'
  } else if (result.awaiting) {
    result.status = 'partial'
    result.reason =
      `Scored ${result.scored} returned submission(s); ${result.awaiting} still out and not ` +
      'counted in any total.'
  }
  return result
}

/** Matches Python's `%g`: 10 stays "10", 9.5 stays "9.5". */
const formatGrade = (grade: number) => String(Number(grade.toPrecision(6)))

const round2 = (value: number) => Math.round(value * 100) / 100
const round1 = (value: number) => Math.round(value * 10) / 10

File: backend/app/api/routes/analytics.py
"""Analytics: ranked opportunities, a selling queue, returns, and saved cuts.

Every endpoint here is a projection of an answer some other engine already gave.
None of them computes a verdict, a value or a cost of its own — if a number
appears below, it came from the decision, market, economics or submission engine
and can be traced back to the card page that shows the same figure.

That is why these are thin. The interesting work happened upstream; this is the
part that decides which of those answers you are asked to look at today.
"""

from __future__ import annotations

from datetime import date
from typing import Annotated

from fastapi import APIRouter, Query, status
from pydantic import Field

from app.api.deps import DbSession
from app.api.errors import ApiError
from app.api.routes.collection import OpportunityOut
from app.enums import Confidence
from app.schemas.common import ApiModel
from app.services import analytics, disposals, portfolio

router = APIRouter(prefix="/analytics", tags=["analytics"])


# --- Opportunities -----------------------------------------------------------


class OpportunitiesOut(ApiModel):
    """Cards worth grading, best first."""

    status: str
    reason: str | None = None
    currency: str = "GBP"
    analysed: int = 0
    total_cards: int = 0
    actionable: int = 0
    expected_profit: float | None = None
    total_grading_cost: float | None = None
    items: list[OpportunityOut] = Field(default_factory=list)


@router.get(
    "/opportunities",
    response_model=OpportunitiesOut,
    summary="Best grading opportunities",
    description=(
        "The same verdicts as `/collection/decisions`, cut down to the cards you would "
        "actually send and ranked by opportunity score.\n\n"
        "Deliberately not a second ranking: two rankings of one question drift apart, and the "
        "one that goes stale is whichever you were not looking at. Every figure here is the "
        "figure that endpoint returns.\n\n"
        "`batch_size` matters — a card that does not pay on its own often pays in a submission "
        "of twenty, so the list changes with the parcel you have in mind."
    ),
)
def opportunities(
    db: DbSession,
    batch_size: Annotated[
        int, Query(ge=1, le=1000, description="Cards assumed to share one submission.")
    ] = 1,
    limit: Annotated[int, Query(ge=1, le=2000)] = portfolio.DEFAULT_LIMIT,
) -> OpportunitiesOut:
    result = analytics.opportunities(db, batch_size=batch_size, limit=limit)
    return OpportunitiesOut(
        status=result.status,
        reason=result.reason,
        currency=result.currency,
        analysed=result.analysed,
        total_cards=result.total_cards,
        actionable=result.actionable,
        expected_profit=result.expected_profit,
        total_grading_cost=result.total_grading_cost,
        items=[OpportunityOut(**vars(item)) for item in result.items],
    )


# --- The selling queue -------------------------------------------------------


class SellingCandidateOut(ApiModel):
    """One card to sell raw, and what to ask for it."""

    card_id: str
    name: str
    set_label: str | None = None
    decision: str
    realistic_sale: float | None = Field(
        default=None, description="What the market says it realistically sells for."
    )
    net_proceeds: float | None = Field(
        default=None, description="What you would keep after fees and postage."
    )
    suggested_listing: float | None = Field(
        default=None, description="What to ask, which is not what it will fetch."
    )
    listing_basis: str | None = None
    liquidity_score: float | None = None
    liquidity_band: str | None = None
    days_since_last_sale: int | None = None
    trend_direction: str | None = None
    confidence: str
    purchase_price: float | None = None
    gain_vs_purchase: float | None = Field(
        default=None, description="Null when you never recorded what you paid — not zero."
    )
    blockers: list[str] = Field(default_factory=list)


class SellingQueueOut(ApiModel):
    status: str
    reason: str | None = None
    currency: str = "GBP"
    analysed: int = 0
    total_cards: int = 0
    total_net_proceeds: float | None = None
    items: list[SellingCandidateOut] = Field(default_factory=list)
    notes: list[str] = Field(default_factory=list)


@router.get(
    "/selling-queue",
    response_model=SellingQueueOut,
    summary="Cards to sell raw, with a price to ask",
    description=(
        "The decision engine says 'sell raw' and stops. This answers the question that comes "
        "next: what to list it at.\n\n"
        "A listing price is a negotiating position, not a valuation. The markup over the "
        "realistic sale price scales with liquidity — a card that trades weekly needs little "
        "room, one that trades twice a year needs plenty — and is capped at the upper quartile "
        "of what people have actually paid, because asking more than that is how a listing sits "
        "unsold.\n\n"
        "`net_proceeds` is the same figure the card page shows, after fees and postage. A card "
        "with no raw sales behind it gets no suggested price and says why."
    ),
)
def selling_queue(
    db: DbSession,
    limit: Annotated[int, Query(ge=1, le=2000)] = portfolio.DEFAULT_LIMIT,
) -> SellingQueueOut:
    result = analytics.selling_queue(db, limit=limit)
    return SellingQueueOut(
        status=result.status,
        reason=result.reason,
        currency=result.currency,
        analysed=result.analysed,
        total_cards=result.total_cards,
        total_net_proceeds=result.total_net_proceeds,
        items=[SellingCandidateOut(**vars(item)) for item in result.items],
        notes=result.notes,
    )


# --- What to assess first ----------------------------------------------------


class AssessmentCandidateOut(ApiModel):
    """One unassessed card, and the most grading could possibly gain."""

    card_id: str
    name: str
    set_label: str | None = None
    verdict: str = Field(description="`assess`, `skip` or `unknown`.")
    reason: str | None = None
    ceiling: float | None = Field(
        default=None,
        description=(
            "The most grading could add, if the card came back at the best-priced grade. "
            "An upper bound, not a forecast — an assessment can only bring it down."
        ),
    )
    ceiling_is_complete: bool = Field(
        default=False,
        description=(
            "False when the best *priced* grade sits below the top of that company's ladder, "
            "which makes the ceiling a bound over the priced grades only."
        ),
    )
    company_code: str | None = None
    tier_name: str | None = None
    grading_cost: float | None = None
    best_grade_label: str | None = None
    best_net: float | None = None
    net_raw_value: float | None = None
    liquidity_score: float | None = None
    liquidity_band: str | None = None
    confidence: str = Confidence.NONE.value


class AssessmentQueueOut(ApiModel):
    status: str
    reason: str | None = None
    currency: str = "GBP"
    analysed: int = 0
    total_cards: int = 0
    unpriced: int = 0
    worth_assessing: int = 0
    ruled_out: int = 0
    unknown: int = 0
    truncated: bool = False
    total_ceiling: float | None = None
    items: list[AssessmentCandidateOut] = Field(default_factory=list)
    notes: list[str] = Field(default_factory=list)


@router.get(
    "/assessment-queue",
    response_model=AssessmentQueueOut,
    summary="Which unassessed cards are worth the five minutes",
    description=(
        "Importing four hundred cards takes a second; assessing four hundred does not. The "
        "decision engine cannot rank them — it needs an assessment to say anything at all — so "
        "this ranks on the one thing already known about every card: what the market pays for it "
        "raw, and what it pays for the same card in a slab.\n\n"
        "The measure is a **ceiling**, not a forecast. Take the best-netting grade that has sales "
        "behind it, subtract what the card already nets raw and what grading would cost, and that "
        "is the most grading could possibly add. A card whose ceiling is negative cannot be worth "
        "grading in any condition, so it is ruled out without ever being looked at.\n\n"
        "**A bound is only a bound over what is priced.** If the best grade with sales behind it "
        "is a 9, a 10 might be worth far more and a negative ceiling proves nothing — those come "
        "back `unknown`, never `skip`. Ceilings are computed within one company, because pairing "
        "one grader's fee with another's slab price describes a route that does not exist."
    ),
)
def assessment_queue(
    db: DbSession,
    batch_size: Annotated[int, Query(ge=1, le=1000)] = 1,
    limit: Annotated[int, Query(ge=1, le=2000)] = portfolio.DEFAULT_LIMIT,
) -> AssessmentQueueOut:
    result = analytics.assessment_queue(db, batch_size=batch_size, limit=limit)
    return AssessmentQueueOut(
        status=result.status,
        reason=result.reason,
        currency=result.currency,
        analysed=result.analysed,
        total_cards=result.total_cards,
        unpriced=result.unpriced,
        worth_assessing=result.worth_assessing,
        ruled_out=result.ruled_out,
        unknown=result.unknown,
        truncated=result.truncated,
        total_ceiling=result.total_ceiling,
        items=[AssessmentCandidateOut(**vars(item)) for item in result.items],
        notes=result.notes,
    )


# --- Submission returns ------------------------------------------------------


class GradedCardOut(ApiModel):
    """One card that came back, and how the prediction held up."""

    card_id: str
    name: str
    predicted_grade: float | None = None
    actual_grade: float | None = None
    surprise: float | None = Field(
        default=None, description="Positive when it graded better than predicted."
    )
    cost: float | None = None
    graded_value: float | None = Field(
        default=None, description="What the slab is worth now, from that grade's own sales."
    )
    net_if_sold: float | None = None
    profit: float | None = None
    blockers: list[str] = Field(default_factory=list)


class SubmissionReturnOut(ApiModel):
    submission_id: str
    reference: str
    company_code: str | None = None
    status: str
    returned_at: str | None = None
    card_count: int = 0
    graded_count: int = 0
    total_cost: float | None = None
    total_value: float | None = None
    total_profit: float | None = None
    roi_pct: float | None = None
    mean_surprise: float | None = Field(
        default=None,
        description=(
            "Mean signed difference between actual and predicted grades. Positive means the "
            "grader was kinder than the model expected."
        ),
    )
    cards: list[GradedCardOut] = Field(default_factory=list)
    status_note: str | None = None


class SubmissionReturnsOut(ApiModel):
    status: str
    reason: str | None = None
    currency: str = "GBP"
    scored: int = 0
    awaiting: int = 0
    total_cost: float | None = None
    total_profit: float | None = None
    roi_pct: float | None = None
    submissions: list[SubmissionReturnOut] = Field(default_factory=list)


@router.get(
    "/submission-returns",
    response_model=SubmissionReturnsOut,
    summary="What the parcels you sent actually returned",
    description=(
        "Predicted grades against the grades that came back, and cost against what the slabs "
        "are now worth.\n\n"
        "Only submissions with grades recorded are scored. Ones still out are counted and "
        "reported separately rather than averaged in at zero, which would make every open "
        "parcel look like a loss.\n\n"
        "A slab whose grade has no sales behind it cannot be valued, and says so instead of "
        "falling back to a neighbouring grade's price."
    ),
)
def submission_returns(db: DbSession) -> SubmissionReturnsOut:
    result = analytics.submission_returns(db)
    return SubmissionReturnsOut(
        status=result.status,
        reason=result.reason,
        currency=result.currency,
        scored=result.scored,
        awaiting=result.awaiting,
        total_cost=result.total_cost,
        total_profit=result.total_profit,
        roi_pct=result.roi_pct,
        submissions=[
            SubmissionReturnOut(
                **{key: value for key, value in vars(entry).items() if key != "cards"},
                cards=[GradedCardOut(**vars(card)) for card in entry.cards],
            )
            for entry in result.submissions
        ],
    )


# --- What you actually made --------------------------------------------------


class DisposalOutcomeOut(ApiModel):
    """One closed position, and what it says about the decision behind it."""

    disposal_id: str
    card_id: str | None = None
    name: str
    sold_on: date
    grade_label: str
    sold_graded: bool
    currency: str = "GBP"
    net_proceeds: float | None = None
    purchase_price: float | None = None
    grading_cost: float | None = None
    realised_profit: float | None = None
    profit_is_complete: bool = False
    market_value_on_the_day: float | None = Field(
        default=None,
        description=(
            "What the card was worth the day it sold, from `price_snapshots`. Today's price "
            "has moved for reasons that have nothing to do with this decision."
        ),
    )
    vs_market_pct: float | None = None
    raw_value_on_the_day: float | None = None
    grading_gain: float | None = Field(
        default=None,
        description="What the slab netted over the raw card that day, less what grading cost.",
    )
    reason: str | None = None


class RealisedOut(ApiModel):
    status: str
    reason: str | None = None
    currency: str = "GBP"
    sold: int = 0
    scored: int = 0
    graded_sales: int = 0
    raw_sales: int = 0
    total_net_proceeds: float | None = None
    total_realised_profit: float | None = None
    total_grading_gain: float | None = None
    items: list[DisposalOutcomeOut] = Field(default_factory=list)
    notes: list[str] = Field(default_factory=list)


@router.get(
    "/realised",
    response_model=RealisedOut,
    summary="What you actually made, against what was predicted",
    description=(
        "Every other figure in this application is a projection. These are the ones that "
        "happened.\n\n"
        "`prediction_results` has scored *grade* predictions since Phase 8, so the app could "
        "tell you it called a PSA 9 correctly while having no idea whether the submission made "
        "money. This is the other half.\n\n"
        "**A profit missing a cost is not reported as a profit.** A sale without a recorded "
        "purchase price, or a graded sale without a recorded grading cost, is counted in the "
        "proceeds and left out of the profit — a total that silently dropped a cost would be "
        "wrong in the flattering direction.\n\n"
        "`market_value_on_the_day` comes from `price_snapshots`, not from today's price: the "
        "question is whether the decision was right when it was made."
    ),
)
def realised(db: DbSession, limit: Annotated[int, Query(ge=1, le=2000)] = 500) -> RealisedOut:
    result = disposals.realised(db, limit=limit)
    return RealisedOut(
        status=result.status,
        reason=result.reason,
        currency=result.currency,
        sold=result.sold,
        scored=result.scored,
        graded_sales=result.graded_sales,
        raw_sales=result.raw_sales,
        total_net_proceeds=result.total_net_proceeds,
        total_realised_profit=result.total_realised_profit,
        total_grading_gain=result.total_grading_gain,
        items=[DisposalOutcomeOut(**vars(item)) for item in result.items],
        notes=result.notes,
    )


# --- Filters -----------------------------------------------------------------


class CollectionFilterOut(ApiModel):
    """A named cut over the collection, defined in terms of engine output."""

    key: str
    label: str
    description: str


class FilterResultOut(ApiModel):
    key: str
    label: str
    description: str
    status: str
    reason: str | None = None
    currency: str = "GBP"
    matched: int = 0
    analysed: int = 0
    total_cards: int = 0
    unclassified: int = Field(
        default=0,
        description=(
            "Cards the engine could not decide, so could not be tested against this cut. "
            "Reported rather than counted as non-matches: unanswered is not 'no'."
        ),
    )
    card_ids: list[str] = Field(default_factory=list)
    items: list[OpportunityOut] = Field(default_factory=list)


@router.get(
    "/filters",
    response_model=list[CollectionFilterOut],
    summary="The saved cuts on offer",
)
def filters() -> list[CollectionFilterOut]:
    return [CollectionFilterOut(**vars(item)) for item in analytics.FILTERS]


@router.get(
    "/filters/{key}",
    response_model=FilterResultOut,
    summary="Apply one saved cut",
    description=(
        "Each filter is a predicate over figures the decision and market engines already "
        "produced — never a fresh definition of the same idea. 'Hard to sell' uses the minimum "
        "liquidity score you configured, so the filter and the verdicts agree.\n\n"
        "Where a figure is unknown the card does not match: an unknown risk is not a low risk. "
        "Those cards are counted in `unclassified` so a short list is never mistaken for a "
        "complete one."
    ),
)
def apply_filter(
    db: DbSession,
    key: str,
    batch_size: Annotated[int, Query(ge=1, le=1000)] = 1,
    limit: Annotated[int, Query(ge=1, le=2000)] = portfolio.DEFAULT_LIMIT,
) -> FilterResultOut:
    try:
        result = analytics.filter_collection(db, key, batch_size=batch_size, limit=limit)
    except ValueError as exc:
        # Not NotFoundError: its message is "X 'y' was not found", which reads as
        # a missing record. The filter set is fixed and known, so the honest
        # message is that this key is not one of them — and it lists them.
        raise ApiError(
            code="not_found",
            message=f"{exc} Available: {', '.join(item.key for item in analytics.FILTERS)}.",
            status_code=status.HTTP_404_NOT_FOUND,
            details={"key": key, "available": [item.key for item in analytics.FILTERS]},
        ) from exc
    return FilterResultOut(
        key=result.key,
        label=result.label,
        description=result.description,
        status=result.status,
        reason=result.reason,
        currency=result.currency,
        matched=result.matched,
        analysed=result.analysed,
        total_cards=result.total_cards,
        unclassified=result.unclassified,
        card_ids=result.card_ids,
        items=[OpportunityOut(**vars(item)) for item in result.items],
    )
